src/path/sentry_path_windows.c

```diff
@@ -164,14 +164,14 @@ sentry__path_join_wstr(const sentry_path_t *base, const wchar_t *other)
         return sentry__path_from_wstr(other);
     } else if (other[0] == L'/' || other[0] == L'\\') {
         if (isalpha(base->path[0]) && base->path[1] == L':') {
-            size_t len = wcslen(other) + 3;
-            sentry_path_t *rv = path_with_len(len);
+            size_t other_len = wcslen(other);
+            sentry_path_t *rv = path_with_len(other_len + 3);
             if (!rv) {
                 return NULL;
             }
             rv->path[0] = base->path[0];
             rv->path[1] = L':';
-            memcpy(rv->path + 2, other, sizeof(wchar_t) * len);
+            memcpy(rv->path + 2, other, sizeof(wchar_t) * other_len);
```

Comment:
The problem is that you copy only `other_len` `wchat_t` over to `rv->path`, and those do not include the terminating `\0`.

Let's assume the following example:

```
other = "/Users"
base = "C:/"
```

In memory, if we assume each cell to be `wchar_t` wide, those (including the expected result) would look like this:

| Variable  | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 
| -------- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| `other`  | `/` | `U` | `s` | `e` | `r` | `s` | `\0` | ☠ | ☠ |
| `base`  | `C` | `:` | `/` | `\0` | ☠  | ☠ | ☠ | ☠ | ☠ |
| `rv->path`  | `C` | `:` | `/` | `U` | `s` | `e` | `r` | `s` | `\0` |

We cannot assume (from within `sentry__path_join_wstr`) that `other` or `base` have allocated memory beyond their `\0` (and yes, that assumes that we received a valid null-terminated string).

In the original code, `len = 6 + 3`, which is the correct size for `rv->path`, but `memcpy` would now access 2 `wchar_t` beyond the null termination of `other`. One immediate fix would be to subtract `2` from `len` in the `memcpy` invocation:

```C
memcpy(rv->path + 2, other, sizeof(wchar_t) * (len - 2));
```

In your fix, `other_len = 6`, `rv->path` is correctly pre-allocated to `other_len + 3`, but now you only copy the contents of the string but not the terminating `\0` from `other`. A fix here would be to add `1` to `other_len` in the `memcpy` invocation or to explicitly set `rv->path[8] = L'\0';` after the copy.

Reply:
Thanks for the visualization! Updated to manually set the null terminator in https://github.com/getsentry/sentry-native/pull/1047/commits/ea858f6872d0c4bd46665c0a2bb04b70164be639 .